aleph/views/bookmarks_api.py

```diff
@@ -1,51 +1,181 @@
 import logging
 from flask import Blueprint, request
-from aleph.views.util import jsonify, require
+from werkzeug.exceptions import NotFound, Forbidden, BadRequest
+from sqlalchemy.dialects.postgresql import insert as postgres_insert
+
+from aleph.core import db
+from aleph.search import DatabaseQueryResult
+from aleph.views.util import jsonify, require, parse_request, get_index_entity
+from aleph.views.serializers import BookmarkSerializer
+from aleph.model import Bookmark
 
 
 log = logging.getLogger(__name__)
 blueprint = Blueprint("bookmarks_api", __name__)
 
 
+@blueprint.route("/api/2/bookmarks", methods=["GET"])
+def index():
+    """Get a list of bookmarks created by the current role
+    ---
+    get:
+      summary: Get bookmarks
+      tags: [Bookmarks]
+      parameters:
+        - in: query
+          name: limit
+          description: Number of bookmarks to return
+          schema:
+            type: number
+        - in: query
+          name: offset
+          description: Number of bookmarks to skip
+      responses:
+        '200':
+          description: OK
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/BookmarksResponse'
+    """
+    authz = request.authz
+    require(authz.logged_in)
+    collection_ids = authz.collections(authz.READ)
+
+    query = Bookmark.query.filter(
+        Bookmark.role_id == request.authz.id,
+        Bookmark.collection_id.in_(collection_ids),
+    ).order_by(Bookmark.created_at.desc())
+    result = DatabaseQueryResult(request, query)
+    return BookmarkSerializer.jsonify_result(result)
+
+
 @blueprint.route("/api/2/bookmarks", methods=["POST"])
 def create():
-    """Creates a new bookmark for the user. This is currently a stub endpoint,
-    as the current version of the bookmarks features stores bookmarks client side.
+    """Bookmark one or more entities.
     ---
     post:
       summary: Create bookmark
+      tags: [Bookmarks]
+      requestBody:
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/BookmarkCreate'
       responses:
         '200':
+          description: OK
           content:
             application/json:
               schema:
-                type: object
-          description: OK
-      tags:
-        - Bookmarks
+                $ref: '#/components/schemas/Bookmark'
     """
-    require(request.authz.logged_in)
-    log.info("User [%s]: Created bookmark", request.authz.role.id)
-    return jsonify({}, 200)
+    require(request.authz.session_write)
+    data = parse_request("BookmarkCreate")
+    entity_id = data.get("entity_id")
+
+    try:
+        entity = get_index_entity(entity_id, request.authz.READ)
+    except (NotFound, Forbidden):
+        raise BadRequest(
+            "Could not bookmark the given entity as the entity does not exist or you do not have access."
```

Comment:
Perhaps: "We're sorry but we are unable to bookmark this entity for you" - But if this happens are we logging/tracking it?